### ISSUE TITLE
Why are "watch" and "unwatch" hidden behind symbols?

### ISSUE THREAD

**Issue by dead-claudia** [proposal-signals/proposal-signals#139]
I understand the hazards for tracking, but I'm curious why these are hidden behind symbols rather than just being exposed as string keys.

- In a UI framework, unwatching ≈ tree disposal, and it could be used as a similar lifecycle hook.
- Explicit watching is almost always for external integration, and this usually finds its way to the user in some way.

Just a design question since I don't really see how this is risky to expose to the end developer.

**Comment by littledan** [comment]
I agree that these APIs are quite fundamental (this is why they are supported!) but everything related to the Watcher is a bit subtle to use correctly, so it is sorted into the subtle namespace. Even within the space of using Watchers, I’ve already seen some misuse of this particular [Signal.subtle.unwatched] API for cases which really should be handled by the caller of Signal.subtle.Watcher.prototype.unwatch, so I see these hooks as actually *more* subtle. Anyway, we can continue iterating on this design for a while, until Stage 2.7.

**Comment by fabiospampinato** [comment]
It's unclear to me if these APIs are more fundamental to support than cleanup functions, as in Solid's `onCleanup` primitive. That could be used to implement these watched/unwatched options, perhaps in a more customizable way also, I think 🤔 Is there an argument for that these options being more fundamental instead?

It may be also worth explicitly saying what these APIs are necessary for.

**Comment by dead-claudia** [comment]
> I agree that these APIs are quite fundamental (this is why they are supported!) but everything related to the Watcher is a bit subtle to use correctly, so it is sorted into the subtle namespace. Even within the space of using Watchers, I’ve already seen some misuse of this particular [Signal.subtle.unwatched] API for cases which really should be handled by the caller of Signal.subtle.Watcher.prototype.unwatch, so I see these hooks as actually *more* subtle. Anyway, we can continue iterating on this design for a while, until Stage 2.7.

@littledan Could you give some examples of this misuse? Only way I could think of is trying to put `unwatch` wrapper logic in the method for a signal used across multiple children.

Worth noting that `watched` and `unwatched` are a bit misleading as hooks. They *sound* like they should always be called on `watcher.watch(...)` and `watcher.unwatch(...)`, and I suspect that *may* be contributing to problems *like* that. Something like `tracked` and `untracked` or something else that doesn't imply "watch" might be better at conveying that intent.

**Comment by littledan** [comment]
@dead-claudia  Yes I agree that the names have been confusing, but I am not sure if “tracked” will imply the right thing about a watcher being there (as opposed to any old computed). Other names considered for the concept are “connected” and “live”.

I feel hesitant to share examples of mistakes (to avoid making the authors feel bad), but it has come up when people try to integrate signals behind their framework.

@fabiospampinato these APIs are somewhat related to but somewhat different from Solid’s onCleanup. Computeds can come back to life just because something which uses them gets watched, which I think doesn’t occur in Solid.

**Comment by fabiospampinato** [comment]
> @fabiospampinato these APIs are somewhat related to but somewhat different from Solid’s onCleanup. Computeds can come back to life just because something which uses them gets watched, which I think doesn’t occur in Solid.

Oh good point, yeah. One can't emulate this with cleanups for computeds that have not been disposed yet but that aren't being pulled anymore. It's an interesting scenario.

**Comment by dead-claudia** [comment]
> I feel hesitant to share examples of mistakes (to avoid making the authors feel bad), but it has come up when people try to integrate signals behind their framework.

@littledan Oh I wasn't asking for specific code, just a general idea of what kinds of anti-patterns/mistakes are being made.

I've lately been stuck *hard* into backend after mainly using Redux in my last job, so I've been a bit insulated from signals use. Just trying to better understand the lay of the land here.

**Comment by littledan** [comment]
The most basic misuse case was using these callbacks when something simpler and more direct would suffice. These are supposed to be for the case when someone else is doing the watching/unwatching.

**Comment by dead-claudia** [comment]
Related: https://github.com/tc39/proposal-signals/issues/178
